### PR TITLE
[libmad] update to 0.16.4

### PR DESCRIPTION
**Pull request by generic-pers0n**
Switches libmad to Tenacity's maintained version, which integrates many patches, including security fixes. Some distros and other projects, notably Fedora and FreeBSD, have already switched to our version. It is fully compatible with libmad 0.15.1b.

Note: the MSVC ARM patch was dropped in this version as I don't believe it's needed anymore. If the patch is still needed, we should proceed with applying it to our fork instead.

(**Disclaimer**: I'm a Tenacity maintainer and also maintain our fork of libmad too)

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [ ] The "supports" clause reflects platforms that may be fixed by this new version
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by generic-pers0n** [comment]
> @generic-pers0n please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

File: ports/libmad/portfile.cmake [hunk 1]
```diff
@@ -2,33 +2,19 @@ if(VCPKG_TARGET_IS_WINDOWS)
...
-vcpkg_from_sourceforge(
+vcpkg_from_git(
     OUT_SOURCE_PATH SOURCE_PATH
-    REPO mad/libmad
-    REF 0.15.1b
-    FILENAME "libmad-0.15.1b.tar.gz"
-    SHA512 2cad30347fb310dc605c46bacd9da117f447a5cabedd8fefdb24ab5de641429e5ec5ce8af7aefa6a75a3f545d3adfa255e3fa0a2d50971f76bc0c4fc0400cc45
-    PATCHES 0001-Fix-MSVC-ARM.patch
+    URL https://codeberg.org/tenacityteam/libmad
+    REF ee7c3cdd4361b7f7c7da52fa3ba9f76cc35230a9
```
Comment (dg0yt) [inline ports/libmad/portfile.cmake:8]: Is there a SHA512-able source? It is needed for asset caching.

File: ports/libmad/portfile.cmake [hunk 2]
```diff
@@ -2,33 +2,19 @@ if(VCPKG_TARGET_IS_WINDOWS)
...
-
-vcpkg_cmake_configure(
-    SOURCE_PATH "${SOURCE_PATH}"
-    OPTIONS
-        "-DVERSION=${VERSION}"
-)
+vcpkg_cmake_configure(SOURCE_PATH "${SOURCE_PATH}")
 
 vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(CONFIG_PATH "lib/cmake/mad")
```
Comment (dg0yt) [inline ports/libmad/portfile.cmake:14]: There is no `PACKAGE_NAME`, but the original `CONFIG_PATH` indicates it should be `mad`. (The default value is the port name.)

File: ports/libmad/portfile.cmake (hunk 1 above)
Comment (generic-pers0n) [reply ports/libmad/portfile.cmake:8]: I tried to get a hash from vcpkg but I wasn't successful.

Comment (generic-pers0n) [reply ports/libmad/portfile.cmake:8]: The source tarball from Codeberg can be used instead of the git repo.

**Comment by JonLiu1993** [comment]
@generic-pers0n, thanks for your PR, this is error log, please take a look:
````
FAILED: mad_example 
: && /android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21 --sysroot=/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/sysroot -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -Werror=format-security  -fPIC   -fno-limit-debug-info -static-libstdc++ -Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments CMakeFiles/mad_example.dir/minimad.c.o -o mad_example  libmad.a  -latomic -lm && :
ld: error: libmad.a(layer3.c.o):(function III_decode: .text.III_decode+0x360): branch and link relocation: R_ARM_THM_CALL to non STT_FUNC symbol: III_imdct_l interworking not performed; consider using directive '.type III_imdct_l, %function' to give symbol type STT_FUNC if interworking between ARM and Thumb is required; layer3.c:2525 (/mnt/vcpkg-ci/b/libmad/src/0-e1f1acd0ee.clean/layer3.c:2525)
ld: error: libmad.a(layer3.c.o):(function III_decode: .text.III_decode+0x49e): branch and link relocation: R_ARM_THM_CALL to non STT_FUNC symbol: III_imdct_l interworking not performed; consider using directive '.type III_imdct_l, %function' to give symbol type STT_FUNC if interworking between ARM and Thumb is required; layer3.c:2550 (/mnt/vcpkg-ci/b/libmad/src/0-e1f1acd0ee.clean/layer3.c:2550)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
````

**Comment by JonLiu1993** [comment]
@generic-pers0n, thanks for the PR, when I installed libmad:x64-windows it looked like it was being used incorrectly:

````
Elapsed time to handle libmad:x64-windows: 11 s
libmad:x64-windows package ABI:
Total install time: 11 s
libmad provides pkg-config modules:

    # MPEG audio decoder library
    mad
````

**Comment by generic-pers0n** [comment]
> @generic-pers0n, thanks for the PR, when I installed libmad:x64-windows it looked like it was being used incorrectly:
> 
> ````
> Elapsed time to handle libmad:x64-windows: 11 s
> libmad:x64-windows package ABI:
> Total install time: 11 s
> libmad provides pkg-config modules:
> 
>     # MPEG audio decoder library
>     mad
> ````

This looks correct. I tested an example on Linux and using the `mad` target works.

This might change upstream later, and if that changes, the usage text will be updated here as needed. For now, this is the correct usage.

File: ports/libmad/portfile.cmake [hunk 3]
```diff
@@ -2,33 +2,26 @@ if(VCPKG_TARGET_IS_WINDOWS)
...
-    OUT_SOURCE_PATH SOURCE_PATH
-    REPO mad/libmad
-    REF 0.15.1b
-    FILENAME "libmad-0.15.1b.tar.gz"
-    SHA512 2cad30347fb310dc605c46bacd9da117f447a5cabedd8fefdb24ab5de641429e5ec5ce8af7aefa6a75a3f545d3adfa255e3fa0a2d50971f76bc0c4fc0400cc45
-    PATCHES 0001-Fix-MSVC-ARM.patch
+vcpkg_download_distfile(
+    ARCHIVE
+    URLS "https://codeberg.org/tenacityteam/libmad/archive/${VERSION}.tar.gz"
+    FILENAME "${VERSION}.tar.gz"
```
Comment (dg0yt) [inline ports/libmad/portfile.cmake:8]:
```suggestion
    FILENAME "tenacityteam-libmad-${VERSION}.tar.gz"
```

**Review by dg0yt — CHANGES_REQUESTED** [review]

**Comment by dg0yt** [comment]
On Linux, I get CMake usage, and this is what we prefer:

~~~
libmad provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(mad CONFIG REQUIRED)
  target_link_libraries(main PRIVATE mad::mad)
~~~

This should also be printed on windows, given the CI file lists.
~~~
libmad:x64-windows:/share/mad/madConfig.cmake
libmad:x64-windows:/share/mad/madConfigVersion.cmake
libmad:x64-windows:/share/mad/madTargets-debug.cmake
libmad:x64-windows:/share/mad/madTargets-release.cmake
libmad:x64-windows:/share/mad/madTargets.cmake
~~~

**Comment by JonLiu1993** [comment]
Displaying it on Windows is more user-friendly, thanks.
````
libmad provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(mad CONFIG REQUIRED)
  target_link_libraries(main PRIVATE mad::mad)
````

**Comment by dg0yt** [comment]
> (**Disclaimer**: I'm a Tenacity maintainer and also maintain our fork of libmad too)

@generic-pers0n Does your git server offer bitwise stable tarballs for tags? If not, `vcpkg_download_distfile` can't be used.  #36362

**Comment by generic-pers0n** [comment]
> > (**Disclaimer**: I'm a Tenacity maintainer and also maintain our fork of libmad too)
> 
> @generic-pers0n Does your git server offer bitwise stable tarballs for tags? If not, `vcpkg_download_distfile` can't be used. #36362

I was under the impression they were bitwise stable, but potentially not. However, Codeberg recently dealt with a massive DDoS attack, so I don't know if that had something to do with causing the hash to change or so.

If needed, we can provide source tarballs ourselves, and I can make one very quickly for libmad's latest release.

**Comment by dg0yt** [comment]
A manual source tarball might be the easiest solution.
Stable SHA512 are necessary for source asset caching. (And in that way, also for offline builds.) vcpkg CI won't notice changes as long as it has a cached artifact, so the bad experience always hits new users. 
(One year ago there was quite some rumour when Github tarballs had changes, and they had to roll back some changes.)

**Comment by generic-pers0n** [comment]
> A manual source tarball might be the easiest solution. Stable SHA512 are necessary for source asset caching. (And in that way, also for offline builds.) vcpkg CI won't notice changes as long as it has a cached artifact, so the bad experience always hits new users. (One year ago there was quite some rumour when Github tarballs had changes, and they had to roll back some changes.)

I will start uploading manual source tarballs soon.

**Comment by generic-pers0n** [comment]
I have uploaded a tarball of the libmad source. Filenames will follow this structure: `libmad-${PORT_VERSION}.tar.gz`

The expected SHA512 hash should now be `5b0a826408395e8b6b8a33953401355d6c2f1b33ec5085530b4ac8a538c39ffa903ce2e6845e9dcad73936933078959960b2f3fbba11ae091fda5bc5ee310df5` for 0.16.4.

**Comment by ryandesign** [comment]
> I was under the impression they were bitwise stable, but potentially not.

https://codeberg.org/Codeberg/Community/issues/1366

> (One year ago there was quite some rumour when Github tarballs had changes, and they had to roll back some changes.)

Yup, that was the [same problem](https://github.blog/2023-02-21-update-on-the-future-stability-of-source-code-archives-and-hashes/). Both caused by updating to git 2.38 and either not noticing that it changed the archive compression method or not understanding the impact that change would have on users.

> A manual source tarball might be the easiest solution.

Yes, all projects everywhere should offer this, for reasons such as this.

**Pull request merged**